Sources/AmplitudeCore/Remote Config/RemoteConfigClient.swift

```diff
@@ -0,0 +1,378 @@
+            for callback in callbacks {
+                switch callback.deliveryMode {
+                case .all:
+                    break
+                case .waitForRemote:
+                    // Wait until the initial callback from registerForUpdates is fired
+                    guard callback.lastCallbackTime == nil else {
+                        continue
+                    }
+                }
+
+                sendCallback(callback, configInfo: updatedRemoteConfig, source: .remote)
+            }
+        }
+    }
+
+    private func sendCallback(_ callbackInfo: CallbackInfo, configInfo: RemoteConfigInfo?, source: Source) {
+        callbackInfo.lastCallbackTime = Date()
+
+        var filteredConfig: RemoteConfig?
+        if let key = callbackInfo.key {
+            filteredConfig = key.split(separator: ".").reduce(configInfo?.config) { config, currentKey in
+                return config?[String(currentKey)] as? RemoteConfig
+            }
+        } else {
+            filteredConfig = configInfo?.config
+        }
+
+        callbackInfo.callback(filteredConfig, source, configInfo?.lastFetch)
+    }
+
+    // MARK: - Fetch
+
+    private func fetch(request: URLRequest? = nil, retries: Int = 3) async throws -> RemoteConfigInfo {
+        let currentRequest = try request ?? makeRequest()
+        let (data, response) = try await urlSession.data(for: currentRequest)
+
+        if let httpResponse = response as? HTTPURLResponse,
+           httpResponse.statusCode == 200,
+           let json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Sendable],
+           let config = json["configs"] as? [String: Sendable] {
+            return RemoteConfigInfo(config: config, lastFetch: Date())
+        } else if retries > 0 {
+            return try await fetch(request: currentRequest, retries: retries - 1)
+        } else {
+            throw RemoteConfigError.badResponse
+        }
+    }
+
+    private func makeRequest() throws -> URLRequest {
+        guard var urlComponents = URLComponents(string: serverUrl) else {
+            throw RemoteConfigError.invalidServerURL
+        }
+
+        var queryItems: [URLQueryItem] = []
+        queryItems.append(URLQueryItem(name: "api_key", value: apiKey))
+
+        for configKey in Config.fetchedKeys {
+            queryItems.append(URLQueryItem(name: "config_keys", value: configKey))
+        }
+
+        urlComponents.queryItems = queryItems
+
+        guard let url = urlComponents.url else {
+            throw RemoteConfigError.invalidServerURL
+        }
+
+        return HttpUtil.makeJsonRequest(url: url)
+    }
+}
+
+final class RemoteConfigUserDefaultsStorage: RemoteConfigClient.Storage, @unchecked Sendable {
+
+    private struct Keys {
+        static let config = "config"
+        static let lastFetch = "lastFetch"
+    }
+
+    private let userDefaults: UserDefaults?
+
+    init(instanceName: String? = nil) {
```

Comment:
Nit: what's the storage key of the ios sdk? I'm asking because browser sdk takes part of api key and doesn't use instance name for storage key.

Reply:
We namespace the container here, either as "com.amplitude.remoteconfig.cache" or "com.amplitude.remoteconfig.cache.instancename". In each container, we have a "config" and "lastFetch" key.

So in a global container, something like `com.amplitude.remoteconfig.cache(.instancename)?.(config|lastFetch)`?

API key would also work, but on key rotation we would drop any existing cached configs, which may not be a big deal.